keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md

```diff
@@ -0,0 +1,616 @@
+
+## Motivation
+
+minReadySeconds specifies the minimum number of seconds for which a newly 
+created Pod should be ready without any of its containers crashing, for it 
+to be considered available. This defaults to 0 (the pod will be considered 
+available as soon as it is ready).
+
+minReadySeconds is already available as an optional field for Deployments,
+DaemonSets. Enabling this option helps in bringing StatefulSets on par 
+with other workload controllers.
+
+### Goals
+
+StatefulSet Controller honoring minReadySeconds and mark Pod ready only if Pod 
+is available for given time mentioned in minReadySeconds.
+
+### Non-Goals
+
+Moving minReadySeconds to Pod Spec is beyond the scope of this KEP 
+because of following reasons:
+- The effort to change pod spec would be large.
+- Currently our workload controllers are inconsistent and we prioritize consistency of experience. 
+- StatefulSets are just different enough from daemonsets and deployments that real world use of minReadySeconds 
+  for stateful sets might influence any future design or point in a more appropriate direction (completeness).
+
+## Proposal
+
+StatefulSet spec should be expanded with an optional field called 
+minReadySeconds. To reflect the StatefulSet pods that honored minReadySeconds,
+an additional field called `AvailableReplicas` will also be added to
+StatefulSet status.
+
+### User Stories (Optional)
+
+#### Story 1
+As a Kubernetes user, I want to ensure that my StatefulSet Pods have the 
+desired state ready before they start serving requests. Some applications 
+should only be considered as available only when the state transfer has 
+happened properly.
+
+#### Story 2
+As a Kubernetes user, I want to leverage StatefulSet's minReadySeconds behind
+service loadbalancers in clouds. This is to prevent killing pods in rotation
+before new pods show up.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+We are proposing a new field called `minReadySeconds` whose default value will 
+be 0. In this mode, StatefulSet will behave exactly like its current behavior.
+Its possible we introduce a bug in the implementation. The mitigation 
+currently is that is disabled by default in Alpha phase behind a feature gate 
+for people to try out and give feedback. In Beta phase when its enabled by 
+default, people will only see issues or bugs when `minReadySeconds` is set to 
+something greater than 0. Since people would have tried this feature in Alpha, we 
+would have had time to fix issues.
+
+## Design Details
+
+### StatefulSet
+
+```go
+type StatefulSetSpec struct {
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	// This is an alpha field and requires enabling StatefulSetMinReadySeconds feature gate.
+	// +optional
+	MinReadySeconds *int32
```

Comment:
For consistency with other controllers this should not be a pointer, especially that 0 and nil have the same meaning.

Reply:
Ya, I was thinking about the same but in the [apiconventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md), it was mentioned that:


- ensure the field is entirely absent from API responses when empty (**optional fields should be pointers, anyway**) which made me think I should use a pointer. It also helps in setting default values easily as we can  distinguish between  being set and the default value.

This was also mentioned in the api conventions doc.
```
Do use pointers to scalars when you need to distinguish between an unset value and an automatic zero value. For example, PodSpec.TerminationGracePeriodSeconds is defined as *int64 the go type definition. A zero value means 0 seconds, and a nil value asks the system to pick a default.
```

I think the above is the case for us too say if we choose to a non-zero default value in future.

Reply:
We spoke with @smarterclayton on slack, and the agreement is that we'll go with non-pointer version.